File: python/atelier_bench/metrics/part_cost.py
"""part-cost metric."""

from __future__ import annotations

from typing import TYPE_CHECKING

from ..contracts.metric_contract import MetricUnit
from ..support.abstract_metric import AbstractMetric

if TYPE_CHECKING:
    from ..dataset import Dataset


class PartCostMetric(AbstractMetric):
    KEY = "part-cost"

    def __init__(self) -> None:
        super().__init__(self.KEY, MetricUnit.CENTS)

    def compute(self, data: "Dataset") -> float:
        return data.parts_cost_cents()

File: go/atelier/metrics/part_cost.go
// Package metrics carries the 16 metrics. Each embeds support.BaseMetric,
// declares only Compute, and registers itself in an init function.
package metrics

import (
	"atelier.example/lane/atelier/contracts"
	"atelier.example/lane/atelier/dataset"
	"atelier.example/lane/atelier/support"
)

// Compile-time proof of satisfaction; nothing else names the contract here.
var _ contracts.MetricContract = (*PartCostMetric)(nil)

// PartCostMetric computes the part-cost metric.
type PartCostMetric struct {
	support.BaseMetric
}

// NewPartCostMetric builds the metric with its key and unit.
func NewPartCostMetric() *PartCostMetric {
	return &PartCostMetric{BaseMetric: support.NewBaseMetric("part-cost", contracts.UnitCents)}
}

// Compute is the only contract method this type declares; Key and Unit are
// promoted from the embedded base.
func (m *PartCostMetric) Compute(data *dataset.Dataset) float64 {
	return float64(data.PartsCostCents())
}

// init registers this metric.
func init() {
	Register(NewPartCostMetric())
}
